File: src/main/java/leet_code/till_april/MergeSortedArray.java
package leet_code.till_april;

import java.util.Arrays;

public class MergeSortedArray {

//    You are given two integer arrays nums1 and nums2, sorted in non-decreasing order, and two integers m and n,
//    representing the number of elements in nums1 and nums2 respectively.
//
//    Merge nums1 and nums2 into a single array sorted in non-decreasing order.

    public static void main(String[] args) {

        int[] arr = new int[]{1, 2, 3, 0, 0, 0};
        merge(arr, 3, new int[]{2, 5, 6}, 3);

        System.out.println(Arrays.toString(arr));
    }


    public static void merge(int[] nums1, int m, int[] nums2, int n) {

        int tail1 = m - 1;
        int tail2 = n - 1;

        int finished =  m  + n - 1;

        while (tail1 >= 0 && tail2 >= 0){
            nums1[finished--] = nums1[tail1] > nums2[tail2] ?
                    nums1[tail1--] : nums2[tail2--];


        }
        while (tail2 >= 0) {
            nums1[finished--] = nums2[tail2--];
        }

    }
}

File: src/main/java/leet_code/top_interview_questions_easy/trees/MergeSortedArray.go
package main

func main() {

}

func merge(nums1 []int, m int, nums2 []int, n int) {

	finalLen := m + n - 1
	midx := m - 1
	nidx := n - 1
	for nidx >= 0 {
		if midx > 0 && nums1[midx] > nums2[nidx] {
			nums1[finalLen] = nums1[midx]
			midx--
		} else {
			nums1[finalLen] = nums2[nidx]
			nidx--
		}
		finalLen--
	}

}
